README
INSTALLATION

Begin by ensuring that (virtually) any version of node is running on your workstation.
The existing version will be used to install Node 6.6.0, the version required for
the CNC server.

  % which node

If node isn't found, install node from your Linux distribution's package manager. On
SuSE, install the 'nodejs' package.  On CentOS:
  # yum install epel-release
  # yum install nodejs
Best to use a terminal emulator; an Emacs bash
buffer doesn't interpret the color highlighting sequences.

Next, install Node Version Manager (nvm).  nvm will download and enable any desired
version of node.  

  % curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.32.0/install.sh | bash

Use nvm to download the required version of node (currently v6.6.0). It will also
configure your PATH variable to find this version of node.

  % nvm install v6.6.0

When the CNC server is built, the scripts assume that the developer's shell
has 'nvm' on the path.  One way to do that is to put something like the
following in your .bashrc:

  export NVM_DIR="/users/nick/.nvm"
  [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
  [[ -r $NVM_DIR/bash_completion ]] && . $NVM_DIR/bash_completion

'node-gyp' is required to be available as a shell command. This installation
must be run as root.  node-gyp is needed to build node.js C++
extension objects.

  # npm install -g node-gyp

The CNC server is an autotools project wrapped around a node.js
project. Configure, build and install is done in the familiar way, except that 
'configure' must be run twice.  The build directory must be called 'Build'
because node_gyp is hardwired to expect stuff there.

  % autoreconf
  % mkdir Build
  % cd Build
  % ../configure --with-pic --with-toolroot=/opt/frqu/TOOLROOT
  % make 

... If the make stops because of a missing config file, run configure/make
again ...

  % ../configure --with-pic --with-toolroot=/opt/frqu/TOOLROOT
  % make 

  % make install
  
The installation process retrieves many of its dependencies from the
Internet, notably everything in node_modules and the Swagger interactive
documentation code.

To use the Swagger API specification editor, invoke:

  % cd Build
  % make swagger-editior
  % cd ../swagger-editor
  % npm start

The cnc_server connects to Redis on startup. See
./installer/vcnc-redis.service.

To run the cnc_server from a development directory:

% cd @p4-workspace-dir@/vcnc-rest
% npm start



ChangeLog
Release 3.4.0

    Conditionally serve the web console based on configuration file.

    Web console port is now offset from the rest port by a configurable
    number.

Release 3.4.0

    Issue the ETIME response to start the asynchronous HTTP protocol
    after 1 second rather than after 10 seconds.

    Recognize EUNATCH as 'Protocol Driver not attached'. It means that
    the PIDL messages are incompatible.

Release 3.3.2

    Simplify description of HTTP errors in the Swagger UI.

    Removed unused files.

    Corrections to the API Overview document.

    Fixed invalid example for creating a workspace.

Release 3.3.1

	Fix broken C++ extension objects.

Release 3.3.0

	Production start-up script bin/vcnc-web

	VELSTOR_VCNC_CONFIG sets the (Pepsis) configuration path.

	Port to node.js v6.6.0

	Enable CORS

Release 3.2.6

	Adapt to new Pepsis libraries.

Release 3.2.5

	Fix for the FRQU library name change.

	All operations are 202/fulfill enabled.

	Issue a JSON-format body with the 202 response in order to keep vCLC happy.

	Make the 202 timeout configurable through vcnc-config.yaml

	Fixed a race condition between the direct and indirect/202 response
	paths.

Release 3.2.4

	Rollback the vcnc script to avoid (perilous) use of npm at runtime.

Release 3.2.3

	Hotfix that generates the swagger.json file at build time.

Release 3.2.2

	Refinements to the production operation scripts.

	Fix a bug where the OpenAPI specification wasn't available at runtime.

Release 3.2.1

	Production-level 202/fulfill implementation.

	Debut of the vcnc configuration file ./config/vcnc-config.yaml

	Moved the Pepsis configuration file to ./config/vcnc-pepsis.conf

Release 3.2.0

	All operations under /vtrq  (currently, all operations) use the
	202/fulfill asynchronous protocol for slow responses.

	202/fulfill is a preview implementation.

	Removed /vpi zero time sync operations.

	Removed /vtrq/.../to_trqspace operation.

Release 3.1.6

	Adapt to new VelStor VPProperties RPC API.

Release 3.1.5

	Port to node v4.4.2

	Remove reliance on any TOOLROOT package other than node itself.

Release 3.1.4

	Adapt to VelStor library name changes

Release 3.1.3

	Meta-data copy uses the more efficient CopyVector RPC.

	Change the default listening port to 5500

	Finish the velstorization of the systemd scripts and documentation

	Eliminate the false error message about "no mentor"

Release 3.1.2
	Adapt the wrapper 'vcnc' script to the new 'stackable'
	installation directory structure.

Release 3.1.1
	Fixed a bug in cnctrqVPFindWorker where an loop index variable was
	not being incremented.

Release 3.1.0
	<B0822> solution provided by updated Pepsis libraries.

Release 3.0.2
	Report and exit if Pepsis configuration file processing fails.

	Remove trailing newlines from the default vcnc.conf file, as a
	work-around to <B0822>.

Release 3.0.1
	Use modified frquWorkspace methods to get the vesltorized JSON syntax.

Release 3.0.0
	New REST API reflects VelStor branding.

Release 2.0.5
	<B4717> Fixed the 'nvm use --delete-prefix' issue.

Release 2.0.4
	Fixed a bug that threw a protocol error when the consistency
	attribute was requested on a non-existant node. <FRQU-86>

	Uses nvm and tools (node-gyp, generate-md) from TOOLROOT.

	Solved a linking issue with a new Pepsis class.

Release 2.0.3
	Re-compiled to match new VelStor binaries.

Release 2.0.2
	Adapted to velstor/vp API change.

Release 2.0.1
	Fixed 'workspace get' of a non-existant workspace.

Release 2.0.0
	Optional recursive delete for namespaces.

Release 1.2.6
	Remove debugging code that broke some operations.

Release 1.2.5
	Bump the minor revision purely to trigger a release rebuild.

Release 1.2.4
	Bump the minor revision purely to trigger a release rebuild.

Release 1.2.3

	Provide an HTTP error code and friendly description for ENOTDIR.

	'mkdir' takes an optional boolean 'parents' query parameter, for
	mkdir -p style directory creation.


Release 1.2.2

	cnc_version --version returns exit code 0, like other *nix
	programs.

	Straighten out swagger-express-middleware chain ordering.


Release 1.2.1

	Fix inability to serve swagger-ui by providing a default
	(non-JSON) HTTP 404 response for a certain URL.

Release 1.2.0

	Synchronous errors and exceptions  will now generate a proper
	JSON response body. Asynchonous errors terminate the server.
	See the comment in 'app.js'

	Malformed URLs will now generate a proper JSON response body.

	Implemented the /_fail/sync URL which generates a synchronous
	error during testing.

	Fixed the backslashification problem with 'clc ws get'

	All operations return a JSON object having 'error_sym' and
	'message' properties.

Release 1.1.0:
	Resolved node.js package deprecation warnings.

	Removed unused packages from package.json.

	Removed extraneous stdout chatter.

	Cleaned up the version/build message printed at startup.

	HTTP listening port optionally specified by environment (VIEWS_CNC_PORT).

	Startup script (bin/cnc) correctly handles symlinks in path.

	Startup script fully exits on error.

	views-cnc.conf has the new TRQ0 specification found in views.conf.

	views-cnc.service sends stdout and stderr to both syslog and the console.

	views-cnc.service default user is 'root'.

	views-cnc.service sets the CNC listening port .

Release 1.0.1:
	TRQ node consistency attribute operations

	TRQ Workspace management operations

Release 1.0.0:
	REST operations required by the PI.

	Fixed <B8886> MetaCopy length > 1

	Fixed <B4279> DeleteNodes length > 1

	Default installation prefix is now /opt/views-cnc


vcnc-rest/app.js
'use strict'; // eslint-disable-line strict
/*
 *    Copyright (C) 2015-2016    IC Manage Inc.
 *
 *    See the file 'COPYING' for license information.
 */
// Set the DEBUG environment variable to enable debug output
// process.env.DEBUG = 'swagger:middleware';

const config = require('./lib/configuration');
const express = require('express');
const Middleware = require('swagger-express-middleware');
const path = require('path');
const http = require('http');
const cors = require('cors');
//
//  Initialize the C++ extension
//
const CnctrqClient = require(path.join(__dirname,
				 'addon/build/Release/cnctrq_client')).CnctrqClient;
const cnctrqClient = new CnctrqClient;
cnctrqClient.call_me_first(__dirname);
const fulfill202 = require('./lib/fulfill202');
/**
 *  Top-level Express application object.
 *  @module
 */
module.exports = (() => {
  //
  // ROUTES ===========================================================

  const app = express();

  //
  //  Configure the static content
  //
  app.use(
    '/v1/doc/api',
    express.static(path.join(__dirname, 'static/swagger-ui/dist'))
  );
  app.use(
    '/v1/doc',
    express.static(path.join(__dirname, 'static/doc-html'))
  );
  app.use(
    '/v1/doc/api/spec',
    express.static(path.join(__dirname, 'api/swagger.json'))
  );

  const middleware = new Middleware(app);

  //
  //  Any app.use(..) outside this block (i.e, lexically after) are,
  //  in effect, /earlier/ in the processing chain, because this
  //  block is a callback whose execution is deferred.
  //
  middleware.init(path.join(__dirname, 'api/swagger.yaml'), () => {
    //
    //  Enable CORS
    //
    app.use(cors({ credentials: true, origin: true }))
    //
    //  Metadata middleware
    //
    app.use(middleware.metadata());
    //
    //  Parse the request, convert to approprate data types.
    //
    app.use(middleware.parseRequest());
    //
    //  Very, very crude URL logging
    //
    app.use((req, res, next) => {
      if ('VELSTOR_VCNC_LOG_REQUESTS' in process.env) {
        console.log(req.method, req.originalUrl); // eslint-disable-line no-console
      }
      next();
    });
    //
    //  Validate the request against the API definition, returning HTTP codes
    //  as appropriate.
    //
    app.use(middleware.validateRequest());
    //
    //  Install routers for each API version and variation
    //
    const v1Router = express.Router(); // eslint-disable-line new-cap
    require(path.join(__dirname, 'routes/v1'))(v1Router);
    app.use('/v1', v1Router);
    //
    //  Install a test router
    //
    const testRouter = express.Router(); // eslint-disable-line new-cap
    require(path.join(__dirname, 'routes/test'))(testRouter);
    app.use('/v1/_test', testRouter);
    //
    // Catch undefined URLs (404) and forward to error handler
    //
    app.use((req, res, next) => {
      if (req.originalUrl === '/v1/doc/api/spec') {
        //
        //  Swagger-ui needs this url to 404 and
        //  also return the default response.
        //
        next();
      } else if (req.originalUrl === '/') {
        res.redirect('/v1/doc');
      } else {
        //
        //  Otherwise, if we go here, we got here, we missed
        //  a URL that was outside the swagger-express-middleware
        //  domain, so perform our uniform JSON response handling.
        //
        const err = new Error('URL not found.');
        err.status = 404;
        next(err);
      }
    });
    // ERROR HANDLING ====================================================
    //
    // This error handling follows the policy that every response body shall
    // contain a JSON object having both an 'error_sym' and a 'message'
    // property.
    //
    // When the server runs in 'development' mode, the object will also
    // have a 'stack' property whose value is the stack trace.
    //
    // The CNC server can throw exception/errors in three different contexts:
    // a) synchronously to Express handling (app.use(..)); b) during an Express
    // callback; and c) within libuv as part of processing a node/V8 extension.
    //
    // The code here handles the ((a)) context.  Handling the ((b)) context
    // is very cumbersome - it involves try/catch within every route handling
    // method.  Case ((c)) is similar to ((b)), only within the C++ extension
    // context.
    //
    // The policy is to not handle the ((b)) and ((c)) cases.  The CNC server
    // will terminate, and in production will be automatically restarted. Testing
    // will have to discover internal bugs which can generate such exceptions.
    // This will be sufficient as long as there is no user (REST) input which can
    // generate such errors.
    //
    // http://stackoverflow.com/a/15913211 is a good explanation of the issues.
    //
    function errorHandler(err, req, res, next, stack) {
      if (err.status === 404) {
        res
        .status(err.status)
        .json({ error_sym: 'ENOENT', message: err.message });
      } else {
        res
        .status(err.status || 500)
        .json({
          error_sym: 'EREMOTEIO',
          message: err.message,
          stack,
        });
        next(err);  // logs this error to the server console
      }
    }

    //
    // During development, a stack trace is sent to the client.
    // An error always causes the server to log the stack to stdout/stderr.
    //
    app.use((err, req, res, next) => {
      errorHandler(
        err,
        req,
        res,
        next,
        app.get('env') === 'development' ? err.stack : undefined
      );
    });


    //
    //  Announce we are ready.
    //
    const port = config.server.port;
    console.log('INFO: accepting REST connections on port', port); // eslint-disable-line
    // create http service
    http.createServer(app).listen(port);
  });

  //
  //  An undocumented error generator for testing.
  //
  app.use('/_fail/sync', () => {
    throw new Error('mock synchronous error for testing');
  });
  //
  //  Install the 202 fulfillment controller
  //
  app.get(`${fulfill202.route}/:id`, (req, res) => {
    // console.log('routed to fulfill ');
    // const result = fulfill202.fetchFulfillmentResource(req.params.id);
    fulfill202.fetchFulfillmentResource(req.params.id)
    .then((result) => {
      // console.log('about to fulfill ', result);
      res.status(200).send(result);
    });
  });

  //
  //  Serve the web admin console on its own port.
  if (config.web.enabled) {
    const admin = express();
    admin.use(
      '/',
      express.static(path.join(__dirname, 'static/vcnc-web'))
    );
    const adminPort = parseInt(config.server.port) + parseInt(config.web.offset);
    console.log('INFO: accepting vcnc-web connections on port', adminPort); // eslint-disable-line
    http.createServer(admin).listen(adminPort);
  }
})();
